src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Extensibility
+{
+    /// <summary>
+    /// Extensions for <see cref="AcquireTokenForClientParameterBuilder"/>
+    /// </summary>
+    public static partial class AbstractConfidentialClientAcquireTokenParameterBuilderExtension
+    {
+        /// <summary>
+        /// Intervenes in the request pipeline, by executing a user provided delegate before MSAL makes the token request. 
+        /// The delegate can modify the request payload by adding or removing  body parameters and headers. <see cref="OnBeforeTokenRequestData"/>
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="builder">The builder to chain options to</param>
+        /// <param name="onBeforeTokenRequestHandler">An async delegate which gets invoked just before MSAL makes a token request</param>
+        /// <returns>The builder to chain other options to.</returns>
+        public static AbstractAcquireTokenParameterBuilder<T> OnBeforeTokenRequest<T>(
```

Comment:
@jmprieur - this is the main extensiblity option, as discussed. The `OnBeforeTokenRequestData` is 2 files down - it's a class made up of: 
```csharp
 /// <summary>
        /// Parameters which will be sent in the request body, as POST parameters.
        /// </summary>
        public IDictionary<string, string> BodyParameters { get; }

        /// <summary>
        /// Headers which will be sent with the request.
        /// </summary>
        public IDictionary<string, string> Headers { get; }

        /// <summary>
        /// The token endpoint, including any query parameters, where the request is being sent to.
        /// </summary>
        public Uri RequestUri { get; }
```